#459842 [C#テキスト]9章演習問題

kurosu-being opened the pull request:
第9章 ファイルの操作

主にFileクラス、FileInfoクラスを使ったファイル操作ついて学習しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-1を実装しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-2を仮実装しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-2を実装しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-1において、ctice9-1-3の内容になっていたものを修正しました。
Practice9-1-3を実装しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-2-を実装しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-2-1以前の問題において、FileInfoクラスではなくFileクラスを使用するように変更しました。
Practice9-3-1を実装しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-4-1を仮装しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-4-1を仮装しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-4-1を仮実装しました。
Practice9-5-1を実装しました

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-5-1のフォーマットを修正しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-1において、CountClass.csを追加しました。
Practice9-1-1において、コメントを修正しました。（同盟→同名）
Practice9-4-1において、不要なコメントを削除しました。
Practice9-5-1において、メソッドにコメントを追加し、問題文のコメントを修正しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-1において、CountClass.csにコメントを追加しました。
Practice9-4-1において、コード整形を行いました。
Practice9-5-1において、コード整形を行いました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-2において、コンソールに出力するコメントを修正しました。
Practice9-1-3において、コンソールに出力するコメントを修正しました。
Practice9-2-1において、コンソールに出力するコメントを修正しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-2-1において、コンソールに出力するコメントを修正しました。(出力されるテキストがoutput.txtであることを明示）

k-miyahara-being commented on 章末問題9/Practice9-1-1/Program9-1-1.cs:
```diff
@@ -0,0 +1,33 @@
+            Console.WriteLine("C#のソースファイルのパスを入力してください(CountClass.csはチケットに貼っています)");
+            string wFilePath = Console.ReadLine();
+
+            if (!File.Exists(wFilePath)) {
+                Console.WriteLine("ファイルが見つかりませんでした。");
+                return;
+            }
```
ファイルがあるかのチェックもしっかり出来ていますね！
どういうバグが起こるか、きちんと想定できているのがすばらしいです！

[追加問題]
拡張子がcs以外の場合は「C#のソースファイルのパスを指定してください」と出力するようにしてみてください。

kurosu-being replied:
ありがとうございます！
これからもどんなバグが起こりうるか検討しながら実装していきたいと思います。

追加問題を実装しました。ご確認ください。

k-miyahara-being commented on 章末問題9/Practice9-1-3/Program9-1-3.cs:
```diff
@@ -0,0 +1,32 @@
+                    string wLine = wReader.ReadLine();
+
+                    int wCount = File.ReadLines(wFilePath, Encoding.UTF8).Count(x => x.Contains("class"));
+                    wClassCount = wCount;
```
wCountは不要ですね。

kurosu-being replied:
ご指摘ありがとうございます。
wCountを削除しました。

k-miyahara-being commented on 章末問題9/Practice9-2-1/Program9-2-1.cs:
```diff
@@ -0,0 +1,30 @@
+            string wInputFileName = wFilePath;
+            string wOutputFileName = "output.txt";
+
+            using (var wStreamWriter = new StreamWriter(wOutputFileName)) {
```
せっかくなら、出力先もユーザーが指定できるとよいですね！
最初に出力先のパスの入力も受け取るように実装してみてください。

今、StreamWriterにファイル名だけ渡していますが、この場合どこに出力されるかは確認できていますか？

kurosu-being replied:
ご指摘ありがとうございます。
ユーザーが出力先のパスを指定できるように修正しました。

また、StreamWriterにファイル名だけ渡している場合にはファイルがDebag配下にくることは確認していたのですが、なぜそこに来るかはきちんと把握できていませんでした。ファイルを読み込んだり書き込んだりする際に、ファイルパスを指定しない場合、そのプログラムが実行されているディレクトリを基準として処理が行われることがあるのですね😀

k-miyahara-being replied:
ばっちりです！👍

k-miyahara-being commented on 章末問題9/Practice9-3-1/Program9-3-1.cs:
```diff
@@ -0,0 +1,34 @@
+            if (wFilePath.Split(',').Length != 2) {
+                Console.WriteLine("正しい数のパスが提供されていません。最初の引数に元のファイル、次の引数に追記先のファイルを指定してください。");
+                return;
+            }
```
良いですね！

書き方を指定するなら、ユーザーにとっては
1つめ入力 → Enter → 2つめ入力 → Enter → 処理スタート
の流れのほうが伝わりやすいかなと思いました。また、それによって2つのパスが確実に入力されるようになります。
バグが起きづらく、ユーザーにとって迷いのない仕様を考えることも大切です。

kurosu-being replied:
ご指摘ありがとうございます！
ユーザーが2回入力する形式に変更しました。
今後の実装においても、ユーザーにとっての動作が明確にすることを意識して実装したいと思います。

k-miyahara-being commented on 章末問題9/Practice9-4-1/Program9-4-1.cs:
```diff
@@ -0,0 +1,45 @@
+    つまり、元のファイル名がGreeting.txtならば、コピー先のファイル名はGreeting_bak.txtという名前にします。コピー先に同名のファイルがある場合は置き換えてください。*/
+    class Program {
+        static void Main(string[] args) {
+            Console.WriteLine("指定したディレクトリ直下にあるファイルを別のディレクトリにコピーします。\nコピーしたいファイルパス,コピー先のディレクトリという形で入力してください。");
```
1つ目は「コピーしたいファイルパス」ではなく「コピーしたいディレクトリパス」ですかね？

kurosu-being replied:
その通りです🙇‍♀️
意図が伝わるよう、にメッセージを修正しました。

k-miyahara-being commented on 章末問題9/Practice9-4-1/Program9-4-1.cs:
```diff
@@ -0,0 +1,45 @@
+                string wFileName = Path.GetFileNameWithoutExtension(wFile);
+                string wExtension = Path.GetExtension(wFile);
+
+                string wNewFileName = $"{wFileName}_Bak{wExtension}";
```
細かいですが、Bは小文字ですね。

kurosu-being replied:
ご指摘ありがとうございます！
_Bak→＿bakに修正しました。

k-miyahara-being commented on 章末問題9/Practice9-4-1/Program9-4-1.cs:
```diff
@@ -0,0 +1,45 @@
+
+                string wNewFilePath = Path.Combine(wTargetDirectory, wNewFileName);
+
+                // コピー先に同名のファイルがなければコピーを実行
+                if (!File.Exists(wNewFilePath)) {
+                    File.Copy(wFile, wNewFilePath);
+                    Console.WriteLine($"{wFile} を {wNewFilePath} にコピーしました。");
+                }
```
問題には「コピー先に同名のファイルがある場合は置き換えてください」とあるので、
このままだと同名のファイルはなにもせず終わってしまいます。

kurosu-being replied:
ご指摘ありがとうございます！
Copy(String, String, Boolean)を用いて、同名のファイルがある場合置き換えるように修正しました。
最初自分で条件分岐を実装しかけたのですが、File.Copyのオーバーロードをふんわりと覚えていて助かりました😂

k-miyahara-being replied:
自前で実装する前に、「こんな機能もあったらいいな」と妄想してメソッドを見直してみると、意外と見つかったりします😁

k-miyahara-being commented on 章末問題9/Practice9-5-1/Program9-5-1.cs:
```diff
@@ -0,0 +1,51 @@
+
+            var wDirectory = new DirectoryInfo(vDirectoryPath);
+
+            FileInfo[] wFiles = wDirectory.GetFiles("*", SearchOption.AllDirectories);
```
変数wFilesはなくてもよいです。
また、この時点で１M以上ファイルだけ絞り込んでおけば、foreachの中でif文を書かなくて済みます。

kurosu-being replied:
ご指摘ありがとうございます！
wFilesを削除し、LINQで1M以上のファイルを絞り込むように変更しました。

yuki-tanaka-being commented on 章末問題9/Practice9-1-1/Program9-1-1.cs:
実行していないので推測で記載します🙏
期待される結果は 1行かと思います。
今の実装だと 4行になっていないでしょうか。
ここでの「classキーワード」の「キーワード」とは、予約語のことを指していると思います。
https://learn.microsoft.com/ja-jp/dotnet/csharp/language-reference/keywords/

ただ、問題文に「リテラル文字列やコメントの中にclassという単語は含まれない」とあるのに、CountClass.cs にリテラル文字列の class があるのが良くわかっていません😅

kurosu-being replied:
ご指摘ありがとうございます！
CountClass.csのリテラル文字列の classを削除し、一行で出力されるように変更しました。
予約済みの識別子のことをキーワードと呼ぶのですね😮一般用語かと思い意味を調べるのを怠っていました。正確に要求にこたえる為には用語の確認や前提の確認が大切ですね。

CountClass.cs は私が独自で作成したものです（キーワードのことを誤解しておりました）。失礼いたしました。

yuki-tanaka-being replied:
>CountClass.cs は私が独自で作成したもの

そういうことでしたか。
それなら、森田さんも指摘されていますがデータがよくなかったですね。
期待するデータとしては、コンストラクタやclassを含む変数やメソッドがあるクラスになると思います。

テストデータを正しく考えられるかも大事な部分になります。
いくら時間をかけてテストをしても、テストケースが不足していたらバグを見逃す可能性があります。
質の高いテストデータを準備できるかは、テスト観点が豊富にあるかに依存するかなと思います。
コツとしては、楽観的ではなく悲観的に考えてみることです。
もしこんなデータがきたら大丈夫だろうか。こんな操作をされたら大丈夫だろうか。などなど。

k-miyahara-being commented on 章末問題9/Practice9-5-1/Program9-5-1.cs:
```diff
@@ -0,0 +1,51 @@
+                Console.WriteLine("1MB以上のファイルは見つかりませんでした。");
+            }
+        }
+
+        /// <summary>
+        /// 1Mバイトのファイル名
+        /// </summary>
+        /// <param name="vDirectoryPath">指定されたディレクトリ</param>
+        /// <returns>ファイル名</returns>
+        static List<string> GetLargeFiles(string vDirectoryPath) {
+            var wLargeFiles = new List<string>();
```
メソッドに分けたのはGoodです！！

変数宣言ごとに空行をあけなくても良いかなーと思います。
空行を入れるタイミングは、処理のかたまりごとをイメージするとよいです。

kurosu-being replied:
ありがとうございます🤗
まだどこまでがひとかたまりの処理なのか掴めてないところもあるのですが、今後読みやすいコードで実装していく上で重要だと思うので意識して行っていきます。

k-miyahara-being reviewed (changes requested):
レビューしました。

yuki-tanaka-being commented on 章末問題9/Practice9-1-3/Program9-1-3.cs:
```diff
@@ -0,0 +1,32 @@
+                    int wCount = File.ReadLines(wFilePath, Encoding.UTF8).Count(x => x.Contains("class"));
+                    wClassCount = wCount;
+                }
+            }
```
コピペミスかな？
StreamReader とそれに関するループ処理は不要ですね。

確認済みかもしれませんが、 FileReadAllLines と File.ReadLines の違いはリファレンスに記載があります。
https://learn.microsoft.com/ja-jp/dotnet/api/system.io.file.readlines?view=netframework-4.7.2
注釈の部分に書かれています。
日本語約がおかしいので英語表示でGoogle翻訳した方がいいかも。↓翻訳バージョン。
![image](https://github.com/kurosu-being/TextC/assets/78195392/c2c22c35-e9fb-4aab-84da-791eb6435dde)

kurosu-being replied:
ご指摘ありがとうございます！
ReadLinesメソッドを使い、不要なループ処理を削除いたしました。
usingの中を書き換えて満足していましたが、ReadLinesメソッドの利点を潰していました。部分的な修正にしないことを今後も特に注意したいと思います。
FileReadAllLines と File.ReadLines の違いはテキストに出てきたのですが、ようやくその違いとReadLinesメソッドの便利さを実感しました😋

yuki-tanaka-being commented on 章末問題9/Practice9-2-1/Program9-2-1.cs:
```diff
@@ -0,0 +1,30 @@
+    　出力するファイル名と同名のファイルがあった場合は、上書きしてください。*/
+    class Program {
+        static void Main(string[] args) {
+            Console.WriteLine("テキストファイル「いろは歌.txt」を読み込み、行の先頭に行番号を振り、その結果をoutput.txtとして別のテキストファイルに出力します。\nパスを入力してください(いろは歌.txtはチケットに貼っています)");
```
改行の文字列を使いたい場合、[`Environment.NewLine`](https://learn.microsoft.com/ja-jp/dotnet/api/system.environment.newline?view=netframework-4.7.2) がおすすめです。
プログラムの実行環境に依存せずに、改行を指定することができます。

kurosu-being replied:
ご指摘ありがとうございます！
{Environment.NewLine}で書き換えました。
使っていこうと思います！

yuki-tanaka-being commented on 章末問題9/Practice9-2-1/Program9-2-1.cs:
```diff
@@ -0,0 +1,30 @@
+            string wInputFileName = wFilePath;
+            string wOutputFileName = "output.txt";
+
+            using (var wStreamWriter = new StreamWriter(wOutputFileName)) {
```
`StreamWriter` を使うなら、`StreamReader` とセットがおすすめです。
一行ずつ読み取って書き込めるので、メモリを無駄に消費しません。
いずれ勉強するであろう非同期処理を使う場合なども相性がいいです。
選択肢があって迷うときは、対になっているようなものは、対になっている者同士で使うのが一番よいです。

kurosu-being replied:
ご指摘ありがとうございます！
StreamReadrで書き換え、修正しました。
読み込み、書き込みでそれぞれ3つずつ選択肢があると思い混乱していたので、対になる処理で覚えたいと思います。

yuki-tanaka-being commented on 章末問題9/Practice9-3-1/Program9-3-1.cs:
```diff
@@ -0,0 +1,34 @@
+            Console.WriteLine("1つ目のテキストファイルの最後に、2つ目のテキストファイルの内容を追加します。\n1つ目のファイルパス,2つ目のファイルパスという書き方で指定してください");
+            string wFilePath = Console.ReadLine();
+
+            if (wFilePath.Split(',').Length != 2) {
+                Console.WriteLine("正しい数のパスが提供されていません。最初の引数に元のファイル、次の引数に追記先のファイルを指定してください。");
+                return;
+            }
+
+            string wSouceFilePath = wFilePath.Split(',')[0];
+            string wAnotherFilePath = wFilePath.Split(',')[1];
```
Split を3回実行していますが、変数に取って使うのがよいです。
単純に無駄な処理をしないためと、バグを埋め込まないためです。
ポイントは、状態を確認したオブジェクトと、それを前提にしたロジックとで、実際に使うオブジェクトが異なっている点です。
11行目で長さ2と確認した配列オブジェクトと、16、17行目でインデクサーでアクセスする配列オブジェクトは異なるものです。
この、確認したものと使うものとで、実際はことなるオブジェクトというのが危険度があがる実装になります。
今は、処理が近いですが、間に別の処理が入って来たりすると、バグを埋め込むリスクがどんどんあがっていきます。

kurosu-being replied:
ご指摘ありがとうございます！
宮原さんにもご指摘ありましたが、この処理であると書き手だけでなくユーザーも不便なコードですね。
また、確認したオブジェクトと使うものとのオブジェクトが異なる、という点も意識しておりませんでした。もっと大きなコードを書くことも見据えて勉強していきます、ありがとうございます！

yuki-tanaka-being commented on 章末問題9/Practice9-3-1/Program9-3-1.cs:
```diff
@@ -0,0 +1,34 @@
+            string wSouceFilePath = wFilePath.Split(',')[0];
+            string wAnotherFilePath = wFilePath.Split(',')[1];
+            if (!File.Exists(wSouceFilePath) || !File.Exists(wAnotherFilePath)) {
+                Console.WriteLine("指定されたファイルが見つかりません。パスに間違いがないかご確認ください。");
```
エラーメッセージのコツですが、問題があったときにユーザーがどうすればよいか分かるのが親切です。
ここで惜しいのは、指定されたパスのどれが間違いなのか、ユーザーは分からない点です。

どうするかは仕様次第ですが、今みたいに短絡評価で行くなら、両方ダメな場合でもまずは1個目がないと返し、1個目が正しくなったら2個目がないと返すか。
それとも、両方チェックして、一回で過不足なく問題をユーザーに返すか。
これはシステムを使う側にとって、どういうものが使いやすいかを考えるポイントになります。

私なら、入力を分けて、それぞれ入力した時点でチェックとかかな。

kurosu-being replied:
ご指摘ありがとうございます！
入力を分け、ユーザーが一つずつ確認できるように修正しました。
エラーメッセージについて、なんとなくテキストの見様見真似でやっていましたが、「問題があったときにユーザーがどうすればよいか分かる」というのが大事な指標のひとつなのですね。今後のコーディングで意識して行っていきたいと思います。

s-morita-being replied:
ちなみに、本格的に実装するなら、エラーは標準エラー出力 (Console.Error) に出力するようにすると、
このプログラムを呼び出す側での使い勝手がよくなります。
```
Program9-3-1 >標準出力.txt 2>エラー出力.txt
```
みたいな。

yuki-tanaka-being commented on 章末問題9/Practice9-3-1/Program9-3-1.cs:
```diff
@@ -0,0 +1,34 @@
+                return;
+            }
+
+            string[] wSourceFileLines = File.ReadAllLines(wSouceFilePath);
+
+            using (StreamWriter wStreamWriter = File.AppendText(wAnotherFilePath)) {
+                foreach (string wLine in wSourceFileLines) {
+                    wStreamWriter.WriteLine(wLine);
+                }
+            }
```
`File.ReadAllLines` を使うなら、書き込みは [`File.WriteAllLines` ](https://learn.microsoft.com/ja-jp/dotnet/api/system.io.file.writealllines?view=netframework-4.7.2#system-io-file-writealllines(system-string-system-collections-generic-ienumerable((system-string)))) がおすすめです。

kurosu-being replied:
ご指摘ありがとうございます！
一行ずつの処理を施したかった為、StreamReaderとStreamWriterを使用し、実装しました。
対になる処理を心がけます。

yuki-tanaka-being commented on 章末問題9/Practice9-3-1/Program9-3-1.cs:
```diff
@@ -0,0 +1,34 @@
+            string wFilePath = Console.ReadLine();
+
+            if (wFilePath.Split(',').Length != 2) {
+                Console.WriteLine("正しい数のパスが提供されていません。最初の引数に元のファイル、次の引数に追記先のファイルを指定してください。");
```
同じファイルパスが指定されたときは問題ないでしょうか？

kurosu-being replied:
ご指摘ありがとうございます！
同じファイルパスが指定された時、エラーメッセージがでるように修正いたしました。

yuki-tanaka-being commented on 章末問題9/Practice9-4-1/Program9-4-1.cs:
```diff
@@ -0,0 +1,45 @@
+
+            string wSourceDirectory = wFilePath.Split(',')[0];
+            string wTargetDirectory = wFilePath.Split(',')[1];
+            if (!Directory.Exists(wSourceDirectory) || !Directory.Exists(wTargetDirectory)) {
```
同じディレクトリが指定された場合は問題ないでしょうか？

kurosu-being replied:
ご指摘ありがとうございます！
同じディレクトリの時、エラーメッセージが表示されるように修正しました。

yuki-tanaka-being commented on 章末問題9/Practice9-4-1/Program9-4-1.cs:
```diff
@@ -0,0 +1,45 @@
+
+                string wNewFilePath = Path.Combine(wTargetDirectory, wNewFileName);
+
+                // コピー先に同名のファイルがなければコピーを実行
+                if (!File.Exists(wNewFilePath)) {
+                    File.Copy(wFile, wNewFilePath);
+                    Console.WriteLine($"{wFile} を {wNewFilePath} にコピーしました。");
+                }
```
問題文は「コピー先に同名のファイルがある場合は置き換えてください。」なので、同名のファイルがあったら上書き（削除して追加など）が期待結果かと思います。

kurosu-being replied:
ご指摘ありがとうございます！
同名のファイルがあるとき、置き換えを行うように修正しました。

yuki-tanaka-being replied:
すでに森田さんが指摘されていますが、惜しいですが上書きコピーになってないです。
コピー先のファイルの有無は確認不要になりますね。

kurosu-being replied:
ご指摘ありがとうございます！
ファイルの有無を確認する条件分岐を削除し、修正を行いました。

yuki-tanaka-being commented on 章末問題9/Practice9-5-1/Program9-5-1.cs:
```diff
@@ -0,0 +1,51 @@
+                return;
+            }
+
+            List<string> wLargeFiles = GetLargeFiles(wFilePath);
```
意味のある塊でロジックをメソッドに分割しているのいいですね👍

yuki-tanaka-being commented on 章末問題9/Practice9-5-1/Program9-5-1.cs:
```diff
@@ -0,0 +1,51 @@
+            FileInfo[] wFiles = wDirectory.GetFiles("*", SearchOption.AllDirectories);
+
+            foreach (FileInfo wFile in wFiles) {
+                if (wFile.Length >= 1024 * 1024) {
```
Byte計算ができていてGood！

yuki-tanaka-being reviewed (commented):
ざっと見てみました。

---

宮原くんのレビューとかぶったー🤣
レビュー内容見ずに書いたので重複しているやつはスルーしてくださいー

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-3において、wCountを削除しました。
Practice9-2-1において、ユーザーがディレクトリを指定するように変更しました。
Practice9-2-1において、ファイルの読み込みをStreamReaderをする方法に変更しました。
Practice9-2-1において、ファイルが出力された時のメッセージを変更しました。
Practice9-4-1において、_Bak→_bakに修正しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-2-1において、ファイルが出力された時のメッセージを変更しました。
Practice9-3-1において、StreamReaderで読み取り、StreamWriterで書き出すように修正しました。また、ユーザーが2回入力を行うように修正しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-4-1において、ユーザーが二回に分けてディレクトリを指定するように変更しました。
Practice9-4-1において、同名のファイルがある際に、上書きするように File.Copy(wFile, wNewFilePath);→ File.Copy(wFile, wNewFilePath, true);に変更しました

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

スタッシュを適用させました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-5-1において、バイト計算の処理をLINQを用いて記述しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-4-1において、不要な空行を削除しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-1において、不要な空行を削除しました。
Practice9-4-1において、不要な空行を削除しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-1において、CountClassのリテラル文字を削除しました。
Practice9-1-1において、ファイル読み込みをReadLinesメソッドに変更しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-1において、不要な行を削除しました。
Practice9-1-3において、usingの処理を削除し,ReadLinesメソッドで書き換えました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-2-1～Practice9-4-1において、改行を{Environment.NewLine}で書き換えました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-4-1において、ディレクトリが同じ時のエラーメッセージを追加しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-1において、追加問題を実装しました。

s-morita-being commented on 章末問題9/Practice9-1-1/Program9-1-1.cs:
```diff
@@ -0,0 +1,40 @@
+                while (!wReader.EndOfStream) {
+                    var wCount = wReader.ReadLine();
+                    string wLine = wReader.ReadLine();
+                    if (wLine.Contains("class")) {
```
問題文通りだとすると、↑は例えば `FindSubclasses` といった識別子もカウントしてしまいますね。
> ・classキーワードの前後には、必ず空白文字がある

kurosu-being replied:
ご指摘ありがとうございます！
正規表現でclassという単語を指定し、「class」という単語のみカウントするように修正しました。

s-morita-being commented on 章末問題9/Practice9-3-1/Program9-3-1.cs:
```diff
@@ -0,0 +1,42 @@
+    class Program {
+        static void Main(string[] args) {
+            Console.WriteLine($"1つ目のテキストファイルの末尾に、2つ目のテキストファイル内容を追加します。{Environment.NewLine}まず追加先のテキストファイルのパスを入力してください");
+            string wSourceFilePath = Console.ReadLine();
```
↓の要件は、コマンドライン**引数**で指定できるように、と読めるのですが、どうでしょう?
> コマンドラインで2つのテキストファイルのパス名を指定できるようにしてください。

kurosu-being replied:
ご指摘ありがとうございます！
ご指摘の通り、当初の問題文の意図はコマンドライン引数で指定することなので、今回の場合入力方法が指定されていますね。
新しい『Practice9-3-1(コマンドライン引数でファイルを指定)』にて、コマンドライン引数で指定するように修正を行いました。

s-morita-being reviewed (changes requested):
研修お疲れ様です。
とりあえず見れたところまでの指摘を送ります。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-1において、追加問題を修正しました。
Practice9-3-1(コマンドライン引数でファイルを指定)を追加し、コマンドライン引数でテキストファイルを指定するコードを追加しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-1において、2重に行を読み込んでいる問題を解消しました。
Practice9-1-1のCountClassにおいて、テスト用にFindSubclassesのオブジェクトを追記しました。

kurosu-being committed:
#463298 [C#テキスト]12章演習問題

Practice9-1-4において、最初に表示されるメッセージを変更しました。

s-morita-being commented on 章末問題9/Practice9-1-1/Program9-1-1.cs:
```diff
@@ -0,0 +1,42 @@
+            using (var wReader = new StreamReader(wFilePath)) {
+                while (!wReader.EndOfStream) {
+                    var wLine = wReader.ReadLine();
+                    var wRegex = new Regex(@"\bclass\b");
```
↓は、正規表現まで使わなくても解けるように、という出題者の配慮だったのかなと思います。
> 前後には、必ず空白文字がある

空白が半角スペースだけでよいのなら  `.Contains(" class ")` (← class の前後に半角スペースあり) にするだけでした。

正規表現の `\b` は単語の境界 boundary の b だと思いますが、単語の境界になるなら空白文字以外にもマッチします。
https://learn.microsoft.com/ja-jp/dotnet/standard/base-types/anchors-in-regular-expressions#word-boundary-b

あと、ループの中で毎回 `new Regex()` を実行するのは無駄が大きいです。
正規表現の初期化にはそれなりの実行コストがかかります。

↓を意識するとよいかと思います。
* ループ内で不変なものは、ループする前に一度だけ初期化する。

性能がよくなりますし、ループ内のコードもすっきりして読みやすくなることも期待できます。

kurosu-being replied:
ご指摘ありがとうございます！
正規表現の初期化をループの外で行うように修正しました。

> ループ内で不変なものは、ループする前に一度だけ初期化する。

何度も同じ処理を実行しないよう、ループの中身は特に気を付ける必要がありますね。
意識ししていこうと思います。

s-morita-being commented on 章末問題9/Practice9-1-2/Program9-1-2.cs:
```diff
@@ -0,0 +1,24 @@
+                return;
+            }
+
+            int wClassCount = File.ReadAllLines(wFilePath, Encoding.UTF8).Count(x => x.Contains("class"));
```
↑の `.Contains("class")` も同様ですね。

kurosu-being replied:
ご指摘ありがとうございます！
( class )でカウントするように修正しました。

s-morita-being commented on 章末問題9/Practice9-1-3/Program9-1-3.cs:
```diff
@@ -0,0 +1,22 @@
+                return;
+            }
+
+            var wCount = File.ReadLines(wFilePath, Encoding.UTF8).Count(x => x.Contains("class"));
```
これもです。

kurosu-being replied:
ご指摘ありがとうございます！
同様に、カウントする文字列を( class )に変更しました。

s-morita-being commented on 章末問題9/Practice9-3-1(コマンドライン引数でファイルを指定)/Program9-3-1Tsuika.cs:
```diff
@@ -0,0 +1,45 @@
+namespace Practice9_3_1_コマンドライン引数でファイルを指定_ {
+    class Program {
+        static void Main(string[] args) {
+            Console.WriteLine($"1つ目のテキストファイルの末尾に、2つ目のテキストファイル内容を追加します。コマンドライン引数で指定してください。");
```
修正しなくてもよいですが、例えば引数が指定されなかった時だけ、使い方の説明を出力してもよいかと思いました。

kurosu-being replied:
ご指摘ありがとうございます！
学習の為、引数が指定されなかった時、使い方の説明を出力する方法を実装してみました。
今後もユーザーの状況を考えてエラー文の表示を変えることを試してみたいと思います。

s-morita-being replied:
修正ありがとうございます。
引数が指定されなかった時だけ、という私の指摘がよくなかったですね。

使う人は、↓の場合にどうすればよいか分からないと思うので、
そういう時こそ正しい使い方を出力してあげると親切かと思います。
```
            if (args.Length != 2) {
                Console.WriteLine("引数が正しくありません。プログラムを終了します。");
                return;
            }
```
引数が指定されなかった時は、↑のケースに包含されるので、いくつも条件分岐を増やす必要もなくなります。
(よくみると `if (args.Length == 0)` の分岐は実行されることがないですね。
 先に必ず `if (args.Length != 2)` の分岐に入ります。)

次の機会にでも考慮してもらえればよいです。

kurosu-being replied:
不要な条件分岐でプログラムが複雑にならないことを前提に、シンプルに情報を伝える必要があるのですね。
今後意識していこうと思います。ご教示ありがとうございます🙇‍♀️

s-morita-being commented on 章末問題9/Practice9-4-1/Program9-4-1.cs:
```diff
@@ -0,0 +1,46 @@
+                string wNewFileName = $"{wFileName}_bak{wExtension}";
+                string wNewFilePath = Path.Combine(wTargetDirectory, wNewFileName);
+
+                if (!File.Exists(wNewFilePath)) {
```
↑だと、コピー先に同名のファイルがない場合のみコピーすることになるので、↓の動作にならないですね。
> コピー先に同名のファイルがある場合は置き換えてください。

kurosu-being replied:
ご指摘ありがとうございます！
ifの条件を削除し、問題の要求に会うように修正しました。

s-morita-being commented on 章末問題9/Practice9-5-1/Program9-5-1.cs:
```diff
@@ -0,0 +1,47 @@
+        /// </summary>
+        /// <param name="vDirectoryPath">指定されたディレクトリ</param>
+        /// <returns>ファイル名</returns>
+        static List<string> GetLargeFiles(string vDirectoryPath) {
```
修正する必要はないですが、再利用可能な汎用的な処理を意識するなら、
1M 固定ではなくて、引数でサイズを指定できるとよいですね。

メソッド名が汎用的なので、それとも合いますし。

kurosu-being replied:
ご指摘ありがとうございます！
確かに、検索したいファイル数ごとにメソッドを用意するような書き方より、引数でサイズ指定が出来た方がよりシンプルで汎用的ですね。
メソッド名も1Mと特定しないものになっていました。メソッドの命名について、より中身の機能を指すように
メソッドを書き終えた後の確認を行いたいと思います。

s-morita-being reviewed (changes requested):
一通り見ました。

yuki-tanaka-being commented on 章末問題9/Practice9-2-1/Program9-2-1.cs:
```diff
@@ -0,0 +1,41 @@
+            string wOutputFilePath = Path.Combine(wOutputDirectory, wOutputFileName);
+
+            using (var wStreamReader = new StreamReader(wFilePath, Encoding.UTF8)) {
+                using (var wStreamWriter = new StreamWriter(wOutputFilePath, false, Encoding.UTF8)) {
```
using を2個以上連続で書く時は、ネスト（段下げ）せずに書くことができちゃいます😆
```csharp
            using (var wStreamReader = new StreamReader(wFilePath, Encoding.UTF8))
            using (var wStreamWriter = new StreamWriter(wOutputFilePath, false, Encoding.UTF8)) {
                // 処理
            }
```

kurosu-being replied:
ご教示ありがとうございます！
知りませんでした😮usingを連続で書くとき使っていこうと思います！

yuki-tanaka-being reviewed (commented):
コメントしました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-1-1~9-1-3において、「class」キーワードの前後に空白があることを考慮して修正しました。
Practice9-2-1において、usingを二回使用していた部分の{}を削除しました。
Practice9-3-1において、コマンドライン引数に何も渡されなかった時、メッセージを表示するようにしました。
Practice9-4-1において、不要なif文を削除しました。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-5-1において、メソッド名をGetLargeFiles→GetFilesMore1Mに変更しました。

s-morita-being commented on 章末問題9/Practice9-3-1(コマンドライン引数でファイルを指定)/Program9-3-1Tsuika.cs:
```diff
@@ -0,0 +1,50 @@
+                return;
+            }
+
+            if (args.Length == 0) {
```
`args.Length == 0` の場合は 10 行目の `args.Length != 2` も成り立つので、 16 行目が実行されることはないですね。
11 行目で使い方まで説明してあげるのが親切かと思います。

kurosu-being replied:
ご指摘ありがとうございます！
args.Length != 2の条件の時点で使い方を説明するように修正しました。
条件分岐をする際に、本当にその条件までたどり着くのかを意識したいと思います。

s-morita-being replied:
修正は不要ですが、可能なら条件分岐が少なくなる選択をしていくのがよいです。
プログラムはすぐ複雑になってしまうので。

kurosu-being replied:
そうなのですね。
今後重なっている条件は極力減らしたいと思います🙇‍♀️

s-morita-being commented on 章末問題9/Practice9-5-1/Program9-5-1.cs:
```diff
@@ -33,7 +33,7 @@ class Program {
         /// <param name="vDirectoryPath">指定されたディレクトリ</param>
         /// <returns>ファイル名</returns>
-        static List<string> GetLargeFiles(string vDirectoryPath) {
+        static List<string> GetFilesMore1M(string vDirectoryPath) {
```
LargerThan1M のがしっくりきますかね。

kurosu-being replied:
ご指摘ありがとうございます！
GetFilesMore1M→LargerThan1M に修正いたしました。

s-morita-being reviewed (changes requested):
@kurosu-being 
修正ありがとうございます。
意図したとおりに動作しないだろう修正があったので Request changes にします。

kurosu-being committed:
#459842 [C#テキスト]9章演習問題

Practice9-5-1において、459842 [C#テキスト]9章演習問題

Practice9-5-1において、メソッド名をGetLargeFiles→GetFilesMore1Mに変更しました。
Practice9-3-1(コマンドライン引数でファイルを指定)において、arg.Length＝＝0の条件分岐を削除し、arg.Length !＝2の時のメッセージを修正しました。

s-morita-being commented on 章末問題9/Practice9-5-1/Program9-5-1.cs:
```diff
@@ -33,7 +33,7 @@ class Program {
         /// <param name="vDirectoryPath">指定されたディレクトリ</param>
         /// <returns>ファイル名</returns>
-        static List<string> GetFilesMore1M(string vDirectoryPath) {
+        static List<string> GetFilesThan1M(string vDirectoryPath) {
```
Larger が抜けてます。

files larger than 1M (1M より大きいファイルの一覧) という意味です。

が、 1M ちょうどを含む「以上」という条件と合ってないですね。
Get1MOrLargerFiles とか。
GetFilesLargerOrEqual1M とか。
GetFilesLargerThanOrEqualTo1M とか。
簡潔で正確な表現にしたいところですが…。

あとはお任せします^^;

kurosu-being replied:
ありがとうございます。以下か未満かなど、重要なことを端的に表すのは難しいですね。
今回はGetFilesLargerOrEqual1Mを使わせていただこうと思います。

s-morita-being reviewed (approved):
@kurosu-being 
修正ありがとうございます。
文言を考え出すとキリがないので、 Approve します。

kurosu-being committed:
459842 [C#テキスト]9章演習問題

Practice9-5-1において、メソッド名をGetFilesMore1M→GetFilesLargerOrEqual1Mに変更しました。

kurosu-being committed:
459842 [C#テキスト]9章演習問題

Practice9-3-1(コマンドライン引数でファイルを指定)において、不要な条件分岐を削除しました。
